src/main/scala/org/broadinstitute/dsde/workbench/sam/service/PolicyEvaluatorService.scala

```diff
@@ -48,6 +58,38 @@ class PolicyEvaluatorService(
     }
   }
 
+  /**
+    * Check if the user has the action on the resource.  A true result means they do, and can be trusted.  A false result only means that the short
+    * circuit approach did not find that to be true, but a more exhaustive check might.
+    *
+    * In many cases users are direct members of policies, and it is very fast to query direct members of a policy (as opposed to
+    * calling isMemberOf for a user), this method leverages that to do a fast check for presence
+    */
+  def hasPermissionShortCircuit(resource: FullyQualifiedResourceId, action: ResourceAction, userId: WorkbenchUserId): IO[Boolean] = {
+
+    // if this is auth domain constrainable (or is unknown), we can not use this approach
+    resourceTypes.get(resource.resourceTypeName) match {
+      case Some(x) if !x.isAuthDomainConstrainable =>
+      // nothing
+      case _ =>
+        return IO { false }
```

Comment:
`return` should never be used in scala. There're a few places where the code wrapping things into IO context, and using `map`, `flatMap` just to get inside the context again. It'll be easier if you just use `for comprehension` instead.

the function can be something like
```
    for {
      res <- resourceTypes.get(resource.resourceTypeName).traverse {
        rt =>
          if (!rt.isAuthDomainConstrainable) {
            // compute the list of roles that contain the action requested
            val roleNamesWithAction = rt.roles.filter(_.actions.contains(action)).map(_.roleName)

            for {
              // get the policies for this resource and filter to ones that contain the roles with the action, or the action directly
              policies <- accessPolicyDAO.listAccessPolicies(resource).map(_.toList.filter( p => p.roles.intersect(roleNamesWithAction).nonEmpty || p.actions.contains( action ) ))
              members = policies.flatMap(_.members)
            } yield members.contains(userId)
          } else IO.pure(false)
      }
    } yield res.getOrElse(false)
```

`.traverse` is very useful, what it does is "do something if the `Option` is defined"...

Reply:
much nicer -- thanks for the feedback